src/inspectors/package/RustPackageInspector.ts

```diff
@@ -0,0 +1,415 @@
+import { PackageInspectorBase } from './PackageInspectorBase';
+import { injectable, inject } from 'inversify';
+import { Types } from '../../types';
+import { IFileInspector } from '../IFileInspector';
+
+import * as TOML from '@iarna/toml';
+import { DependencyType, Package, PackageVersion } from '../IPackageInspector';
+
+const isRecord = function (v: unknown): v is Record<string, unknown> {
+  return typeof v === 'object' && v !== null;
+};
+const isString = function (v: unknown): v is string {
+  return typeof v === 'string';
+};
+const isOptString = function (v: unknown): v is string | undefined {
+  return typeof v === 'string' || v === undefined;
+};
+const isOptBool = function (v: unknown): v is boolean | undefined {
+  return typeof v === 'boolean' || v === undefined;
+};
+const isOptStringArray = function (v: unknown): v is string[] | undefined {
+  return v === undefined || (Array.isArray(v) && v.every((e) => typeof e === 'string'));
+};
+
+/**
+ * Similar to `lodash.conformsTo` but runs the check on `undefined` instead of failing.
+ */
+const conformsOptional = function (check: Record<string, (v: unknown) => boolean>, value: Record<string, unknown>): boolean {
+  return Object.keys(check).every((key) => check[key](value[key]));
+};
+
+@injectable()
+export class RustPackageInspector extends PackageInspectorBase {
+  private readonly fileInspector: IFileInspector;
+  cargoLock?: CargoLock;
+  cargoManifest?: CargoManifest | CargoWorkspaceManifest;
+
+  constructor(@inject(Types.IFileInspector) fileInspector: IFileInspector) {
+    super();
+
+    this.fileInspector = fileInspector;
+  }
+
+  async init(): Promise<void> {
+    try {
+      this.debug('RustPackageInspector init started');
+
+      const cargoLockString = await this.fileInspector
+        .readFile('Cargo.lock')
+        .catch((err) => (err.code === 'ENOENT' ? Promise.resolve(undefined) : Promise.reject(err)));
```

Comment:
It doesn't actually have different files with packages, I'm simply reading the lock file so I can correctly fill out the `lockfileVersion` in `Package` type.

In my opinion the question here comes down to whether we are willing to tolerate the additional IO operation when calling `.exists`. The pattern you are suggesting comes down to calling the `lstat` and `readdir` POSIX api functions back to back. The current solution only calls `readdir` and reacts to its error. I thus believe that the current solution is more performant. However, I do agree about readability suffering somewhat. 

I dare suggest two alternatives from which even the `JavaPackageInspector` might benefit in readability:

1. Define the following adaptor on `IFileInspector` (and implement it on `FileInspector`):
```js
static allowNotFound<T>(p: Promise<T>): Promise<T | undefined> {
    return p.catch((err) => (err.code === 'ENOENT' ? Promise.resolve(undefined) : Promise.reject(err)))
}
```
and call it like `FileInspector.allowNotFound(fileInspector.readFile(<args>))`.

2. On the same interface, define more lenient function for `readFile` and `readDirectory` in pattern of:
```js
readFileOptional(path: string): Promise<string | undefined> {
    this.readFile(path).catch((err) => err.code === 'ENOENT' ? Promise.resolve(undefined) : Promise.reject(err))
}
```
and then check for `undefined`.

Whether or not you are willing to accept this is entirely up to you of course, but I do believe it is a better fix for this missing functionality.

I'd also like to point out that I didn't come up with this on my own, I simply saw it in the [implementation of](https://github.com/DXHeroes/dx-scanner/blob/bf3cfacdcd149f16f591269806f275ce70e2ba3b/src/inspectors/FileInspector.ts#L114) `scanFor` function because I originally wanted to use it instead of `readDirectory` in the component detector.